backend/x11/backend.c

```diff
@@ -0,0 +1,374 @@
+	case XCB_BUTTON_INDEX_3: return BTN_RIGHT;
+	// XXX: I'm not sure the scroll-wheel direction is right
+	case XCB_BUTTON_INDEX_4: return BTN_GEAR_UP;
+	case XCB_BUTTON_INDEX_5: return BTN_GEAR_DOWN;
+	default: return 0;
+	}
+}
+
+static bool handle_x11_event(struct wlr_x11_backend *x11, xcb_generic_event_t *event) {
+	struct wlr_x11_output *output = &x11->output;
+
+	switch (event->response_type) {
+	case XCB_EXPOSE: {
+		wl_signal_emit(&output->wlr_output.events.frame, output);
+		break;
+		xcb_key_press_event_t *ev = (xcb_key_press_event_t *)event;
+		struct wlr_event_keyboard_key key = {
+			.time_sec = ev->time / 1000,
+			.time_usec = (ev->time % 1000) * 1000,
+			.keycode = ev->detail - 8,
+			.state = WLR_KEY_PRESSED,
+		};
+
+		wl_signal_emit(&x11->keyboard.events.key, &key);
+		break;
+	}
+	case XCB_KEY_PRESS:
+	case XCB_KEY_RELEASE: {
+		xcb_key_press_event_t *ev = (xcb_key_press_event_t *)event;
+		struct wlr_event_keyboard_key key = {
+			.time_sec = ev->time / 1000,
+			.time_usec = (ev->time % 1000) * 1000,
+			.keycode = ev->detail - 8,
+			.state = event->response_type == XCB_KEY_PRESS ?
+				WLR_KEY_PRESSED : WLR_KEY_RELEASED,
+		};
+
+		wl_signal_emit(&x11->keyboard.events.key, &key);
+		break;
+	}
+	case XCB_BUTTON_PRESS:
+	case XCB_BUTTON_RELEASE: {
+		xcb_button_press_event_t *ev = (xcb_button_press_event_t *)event;
+		struct wlr_event_pointer_button button = {
+			.device = &x11->pointer_dev,
+			.time_sec = ev->time / 1000,
+			.time_usec = (ev->time % 1000) * 1000,
+			.button = xcb_button_to_wl(ev->detail),
+			.state = event->response_type == XCB_BUTTON_PRESS ?
+				WLR_BUTTON_PRESSED : WLR_BUTTON_RELEASED,
+		};
+
+		wl_signal_emit(&x11->pointer.events.button, &button);
+		break;
+	}
+	case XCB_MOTION_NOTIFY: {
+		xcb_motion_notify_event_t *ev = (xcb_motion_notify_event_t *)event;
+		struct wlr_event_pointer_motion_absolute abs = {
+			.device = &x11->pointer_dev,
+			.time_sec = ev->time / 1000,
+			.time_usec = (ev->time % 1000) * 1000,
+			.x_mm = ev->event_x,
+			.y_mm = ev->event_y,
+			.width_mm = output->wlr_output.width,
+			.height_mm = output->wlr_output.height,
+		};
+
+		wl_signal_emit(&x11->pointer.events.motion_absolute, &abs);
+		break;
+	}
+	case XCB_CONFIGURE_NOTIFY: {
+		xcb_configure_notify_event_t *ev = (xcb_configure_notify_event_t *)event;
+
+		output->wlr_output.width = ev->width;
+		output->wlr_output.height = ev->height;
+		wlr_output_update_matrix(&output->wlr_output);
+		wl_signal_emit(&output->wlr_output.events.resolution, output);
+
+		// Move the pointer to its new location
+		xcb_query_pointer_cookie_t cookie =
+			xcb_query_pointer(x11->xcb_conn, output->win);
+		xcb_query_pointer_reply_t *pointer =
+			xcb_query_pointer_reply(x11->xcb_conn, cookie, NULL);
+		if (!pointer) {
+			break;
+		}
+
+		struct wlr_event_pointer_motion_absolute abs = {
+			.device = &x11->pointer_dev,
+			//.time_sec = ev->time / 1000,
+			//.time_usec = (ev->time % 1000) * 1000,
```

Comment:
Commented out?

Reply:
If you don't have anything from x11 here, just grab the current monotonic time.

Reply:
I'm not sure what the x server bases its time off of. It would probably be better to deliver time from a single source. so I guess we should replace all of the events with monotonic time.